text/0000-forget-marker-trait.md

```diff
@@ -0,0 +1,803 @@
+
+This new auto trait is added to the `core::marker` and `std::marker` modules:
+
+```rust
+pub unsafe auto trait Forget { }
+```
+
+Let `T` be `T: !Forget` and `value` be a value borrowed by value of type `T`. Unsafe code is given the following guarantees:
+
+- If `value` is borrowed by `T` as `&mut`, `value` cannot be moved/invalidated/borrowed until `T` is dropped.
+- If `value` is borrowed by `T` as `&`, `value` cannot be moved/invalidated/exclusively borrowed until `T` is dropped.
+
+In practice, we disallow skipping the destructor of `!Forget` types before they exit the scope. Violation is not an immediate undefined behavior, but other code can rely on the destructor running, which can lead to undefined behavior down the road. Unsafe code authors can freely violate this rule, if responsibility is taken.
+
+Several observations can be made about this guarantee. For `T: 'static` we don't have to run the destructor to fulfill it, as `T: 'static` can only have `'static` borrows, which are assumed to be valid indefinite borrows (like with [`Pin::static_ref`]). Another one is, memory borrowed by  `T: !Forget` type cannot be reused or invalidated, as safe code needs to move/take a reference to `value`, similar to the [drop guarantee]. `value` cannot be dropped, as it requires moving.
+
+```rust
+struct Foo<T>(PhantomNonForget, T);
+struct Baz;
+
+impl<T> Drop for Foo<T> { fn drop(&mut self) { } }
+
+let ref_buf = [0u8; 64];
+let mut_buf = [0u8; 64];
+
+// We have a guarantee, that no `&mut` can be taken to `ref_buf` until `Foo`'s `drop`.
+let foo_ref = Foo(PhantomNonForget, &ref_buf);
+// We have a guarantee, that no `&/&mut` can be taken to `ref_buf` until `Foo`'s `drop`.
+let foo_mut = Foo(PhantomNonForget, &mut ref_buf);
+
+drop(ref_buf); // error[E0505]: cannot move out of `ref_buf` because it is borrowed
+drop(mut_buf); // error[E0505]: cannot move out of `mut_buf` because it is borrowed
+
+let ref_first_byte = ref_buf.0[0]; // Allowed
+let mut_first_byte = mut_buf.0[0]; // error[E0503]: cannot use `mut_buf.0[_]` because it was mutably borrowed
+
+drop(foo_ref);
+drop(foo_mut);
+
+let mut_first_byte = mut_buf.0[0]; // Allowed
+drop(ref_buf); // Allowed
+drop(mut_buf); // Allowed
+
+// `Baz` cannot be moved or exclusively borrowed until `Foo` is dropped.
+fn phantom<'a>(baz: &'a Baz) -> Foo<PhantomData<&'a ()>> {
+    Foo(PhantomNonForget, PhantomData)
+}
+```
+
+[`Pin::static_ref`]: https://doc.rust-lang.org/std/pin/struct.Pin.html#method.static_ref
+
+Type becomes `!Forget` if it directly contains `!Forget` member.
+
+We should either allow `!Forget` types in statics or make all `'static` types `Forget` because it fulfills the unsafe guarantee and we can't enforce any code running before the program's abortion.
+
+```rust
+let mut resource = [0u8; 64];
+let _unforget = Subsystem::execute(&mut resource);
+std::process::abort(); // `resource` is (forcefully) borrowed for `'static`
+resource[0] = 42; // unreachable
+```
+
+## Standard Library
```

Comment:
Would this and following sections make more sense as part of the migration section?
It seems we start talking about migration here.